### PR TITLE
log_vc_info: handle long command output

### PR DESCRIPTION
**Pull request by oliver-sanders**
* Log the commands run by log_vc_info (debug level).
* Handle the risk of large command output clogging up the buffer causing commands to hang.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: cylc/flow/pipe_poller.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        # read any data from files
+        nonlocal chunk_size, files
+        for file in select(list(files), [], [], timeout)[0]:
+            buffer = file.read(chunk_size)
+            if len(buffer) > 0:
+                _files[file] += buffer
+
+    while proc.returncode is None:
+        # poll the process (updates the returncode)
+        proc.poll()
```
Comment (MetRonnie) [inline cylc/flow/pipe_poller.py:69]:
```suggestion
    while proc.poll() is None:
```

File: cylc/flow/pipe_poller.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+def pipe_poller(proc, *files, chunk_size=4096):
+    """Read from a process without hitting buffer issues.
+
+    Standin for subprocess.Popen.communicate.
+
+    When PIPE'ing from subprocesses, the output goes into a buffer. If the
+    buffer gets full, the subprocess will hang trying to write to it.
+
+    This function polls the process, reading output from the buffers into
+    memory to prevent them from filling up.
```
Comment (MetRonnie) [inline cylc/flow/pipe_poller.py:38]: In what cases might you call this function that isn't the case of `pipe_poller(proc, proc.stdout, proc.stderr)`? I.e. when would `files` be something other than the output of `proc`?

**Review by MetRonnie — APPROVED** [review]

**Review by MetRonnie — COMMENTED** [review]
(Oh, still need changelog)

**Review by MetRonnie — APPROVED** [review]

File: cylc/flow/install_plugins/log_vc_info.py [hunk 3]
```diff
@@ -63,12 +63,22 @@
...
-    Any, Dict, Iterable, List, Optional, TYPE_CHECKING, TextIO, Union, overload
+    Any,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    TYPE_CHECKING,
+    TextIO,
+    Union,
+    overload,
```
Comment (wxtim) [inline cylc/flow/install_plugins/log_vc_info.py:74]: Pedantic
```suggestion
    TYPE_CHECKING,
    Any,
    Dict,
    Iterable,
    List,
    Optional,
    TextIO,
    Union,
    overload,
```

File: cylc/flow/pipe_poller.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Utility for preventing pipes from getting clogged up.
+
+If you're reading files from Popen (i.e. to extract command output) where the
+command output has the potential to be long-ish, then you should use this
+function to protect against the buffer filling up.
+
+Note, these is a more advanced version of this baked into the subprocpool.
```
Comment (wxtim) [inline cylc/flow/pipe_poller.py:23]: I have no idea which is the more advanced version from this.

```suggestion
Note, there is a more advanced version of this baked into the subprocpool.
```
or
```suggestion
Note, this is a more advanced version of a pipe_poller baked into the subprocpool.
```

**Review by wxtim — COMMENTED** [review]
Seems generally sane. 

I wonder if it might be good to have a functional test which uses bash `timeout` to demonstrate that this approach is safer (i.e. to get an actual script to hang with proc.communicate).

**Comment by oliver-sanders** [comment]
I don't think using a Bash timeout will replicate the problem being fixed here?

**Review by MetRonnie — APPROVED** [review]

**Review by wxtim — APPROVED** [review]

**Comment by wxtim** [comment]
Have approved - the MacOs Tests look to be failing in a standard way. I've kicked flaky and the tutorial tests just to be sure.

**Comment by oliver-sanders** [comment]
The tutorials tests are failing here as on master at the install step.

**Pull request merged**